os/kernel/wqueue/kwork_slmthread.c

```diff
@@ -0,0 +1,238 @@
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/* The state of the kernel mode, high priority softlockup monitor work queue. */
+
+struct hp_wqueue_s g_softlockup_monitor_hpwork;
+
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: softlockup_monitor_work
+ *
+ * Description:
+ *   This work function will be called for pre-defined regular interval.
+ *   It does following operations
+ *   1) It will check the number of context switch operations occured since
+ *      it was called the last time
+ *   2) Whether is there any other thread running except idle thread
+ *   3) Based on above information, it will decide which thread is hogging
+ *      CPU and calls PANIC()
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void softlockup_monitor_work(FAR void *arg)
+{
+		/*Check for the thread next to line in ready to tun list */
+	FAR struct tcb_s *rtcb = (FAR struct tcb_s *)g_readytorun.head;
+	FAR struct tcb_s *ntcb = rtcb->flink;
+	struct work_s *test_wq;
+
+	if (context_switch_counter < 2 && ntcb && ntcb->pid != 0)
+	{
+#if (CONFIG_TASK_NAME_SIZE > 0)
+		lldbg("Lockup caused by [%s] and PID is %d\n", ntcb->name, ntcb->pid);
+#endif
+		PANIC();
+	}
+	/* Reset the counter */
+	context_switch_counter = 0;
+
+	/* Queue the worker function again */
+	test_wq = (struct work_s *) malloc(sizeof(struct work_s));
+	work_queue(SLMWORK, test_wq, softlockup_monitor_work, NULL, CONFIG_SOFT_LOCK_MONITOR_PERIOD);
+}
+
+/****************************************************************************
+ * Name:  work_softlockup_monitor_hpthread
+ *
+ * Description:
+ *   This is the worker thread that performs the actions placed on the high
+ *   priority softlockup_monitor work queue.
+ *
+ * Input parameters:
+ *   argc, argv (not used)
+ *
+ * Returned Value:
+ *   Does not return
+ *
+ ****************************************************************************/
+
+static int work_softlockup_monitor_hpthread(int argc, char *argv[])
+{
+	/* Loop forever */
+
+	for (;;) {
+		/* Then process queued work. work_process will not return until
+		 * 1) There is no further work in the work queue and
+		 * 2) the polling period provided by g_softlockup_monitor_hpwork.delay
+		 *    expires.
+		 */
+
+		work_process((FAR struct kwork_wqueue_s *)&g_softlockup_monitor_hpwork, g_softlockup_monitor_hpwork.delay, 0);
```

Comment:
Why don't you use hpwork? Why do you make new wq?

Reply:
HPWORK is meant for handling the high priority worker functions if any. However if a worker function inside HPWORK gets hogged, Our Monitor worker function couldn't be scheduled inside HPWORK. Since We need a dedicated , reliable workqueue which can monitor the complete system, am using separate workqueue whose priority is maximum ( 255 )